sn_client/src/wallet.rs

```diff
@@ -104,49 +94,67 @@ impl WalletClient {
     /// This can optionally verify the store has been successful (this will attempt to GET the dbc from the network)
     pub async fn pay_for_storage(
         &mut self,
-        content_addrs: impl Iterator<Item = &XorName>,
+        content_addrs: impl Iterator<Item = NetworkAddress>,
         verify_store: bool,
-    ) -> Result<(PaymentProofsMap, Option<Token>)> {
+    ) -> Result<(PaymentTransactionsMap, Option<Token>)> {
         // Let's filter the content addresses we hold payment proofs for, i.e. avoid
         // paying for those chunks we've already paid for with this wallet.
-        let mut proofs = PaymentProofsMap::default();
-        let addrs_to_pay: Vec<&XorName> = content_addrs
+        let mut proofs = PaymentTransactionsMap::default();
+        let addrs_to_pay: Vec<NetworkAddress> = content_addrs
             .filter(|name| {
                 if let Some(proof) = self.wallet.get_payment_proof(name) {
-                    proofs.insert(**name, proof.clone());
+                    proofs.insert(name.clone(), proof.clone());
                     false
                 } else {
                     true
                 }
             })
             .collect();
 
-        let number_of_records_to_pay = addrs_to_pay.len() as u64;
+        let _number_of_records_to_pay = addrs_to_pay.len() as u64;
         // If no addresses need to be paid for, we don't have to go further
         if addrs_to_pay.is_empty() {
             trace!("We already hold payment proofs for all the Chunks.");
             return Ok((proofs, None));
         }
 
-        // Let's build the payment proofs for list of content addresses
-        let (root_hash, audit_trail_info) = build_payment_proofs(addrs_to_pay.into_iter())?;
-        let num_of_addrs = audit_trail_info.len() as u64;
+        let mut total_cost = Token::zero();
 
-        // Always check storage cost, and overpay to allow margin when validation.
-        self.set_store_cost_from_random_address().await?;
-        let storage_cost = Token::from_nano(
-            self.store_cost().as_nano() * (2.0f64.powf((num_of_addrs / 100 + 1) as f64)) as u64,
-        );
+        for content_addr in addrs_to_pay {
+            // TODO: parallelise this
+            let (payments, cost) = self
+                .pay_for_storage_at_address(&content_addr, verify_store)
+                .await?;
 
-        info!("Storage cost per record: {}", storage_cost);
+            if let Some(cost) = total_cost.checked_add(cost) {
+                total_cost = cost;
+            }
 
-        let amount_to_pay = number_of_records_to_pay * storage_cost.as_nano();
+            // store the payment proof
+            proofs.insert(content_addr.clone(), payments);
+        }
 
-        trace!("Making payment for {num_of_addrs} addresses of {amount_to_pay:?} nano tokens.");
+        Ok((proofs, Some(total_cost)))
```

Comment:
same as above, we don't care about the `total_cost`, unless for some logging purpose ?

Reply:
I think for gauging the total cost of a file we may want this?